src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,188 @@
+            if (_consoleBuffer.Writer.TryWrite(evt) == false)
+            {
+                Console.WriteLine(evt);
+            }
+        }
+
+        internal void StartProcessingBuffer(bool batched)
+        {
+            // intentional no-op if the task is already running
+            if (_consoleBufferReadLoop == null || _consoleBufferReadLoop.IsCompleted)
+            {
+                _consoleBufferReadLoop = ProcessConsoleBufferAsync(batched);
+            }
+        }
+
+        private async Task ProcessConsoleBufferAsync(bool batched)
+        {
+            try
+            {
+                if (batched)
+                {
+                    await ProcessConsoleBufferBatchedAsync();
+                }
+                else
+                {
+                    await ProcessConsoleBufferNonBatchedAsync();
+                }
+            }
+            catch (Exception ex)
+            {
+                // Not sure what can fail here, but there is nothing else monitoring this task so just log the exception
+                _exceptionhandler(ex);
+            }
+            finally
+            {
+                // if this has failed for any reason, fall everything back to console
+                _writeEvent = Console.WriteLine;
+                _consoleBuffer.Writer.TryComplete();
+            }
+        }
+
+        private async Task ProcessConsoleBufferNonBatchedAsync()
+        {
+            await foreach (var line in _consoleBuffer.Reader.ReadAllAsync())
+            {
+                Console.WriteLine(line);
+            }
+        }
+
+        private async Task ProcessConsoleBufferBatchedAsync()
+        {
+            var builder = new StringBuilder();
+
+            while (await _consoleBuffer.Reader.WaitToReadAsync())
+            {
+                if (_consoleBuffer.Reader.TryRead(out string line1))
+                {
+                    // Can we synchronously read multiple lines?
+                    // If yes, use the string builder to batch them together into a single write
+                    // If no, just write the single line without using the builder;
+                    if (_consoleBuffer.Reader.TryRead(out string line2))
+                    {
+                        builder.AppendLine(line1);
+                        builder.AppendLine(line2);
+                        int lines = 2;
+
+                        while (lines < DefaultBufferSize && _consoleBuffer.Reader.TryRead(out string nextLine))
+                        {
+                            builder.AppendLine(nextLine);
+                            lines++;
+                        }
+
+                        Console.Write(builder.ToString());
+                        builder.Clear();
+                    }
+                    else
+                    {
+                        Console.WriteLine(line1);
+                    }
+                }
+            }
+        }
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
Do you intend for this class to be inherited? If not, you can ignore the dispose(true/false) pattern by adding `sealed` to your class declaration.